float32[]   offset
---
float32[]   angles
float32[]   xyz
bool        collision
bool        done
